python/cudf/cudf/utils/_numba_setup.py

```diff
@@ -0,0 +1,148 @@
+
+
+def _setup_numba():
+    """
+    Configure numba for use with cuDF. This consists of potentially
+    putting numba into enhanced compatibility mode based on the user
+    driver and runtime versions as well as the version of the cuda
+    toolkit used to build the PTX files shipped with the user cuDF
+    package. It also sets any other config options within numba that
+    are desired for cuDF's operation.
+    """
+    _setup_numba_linker(ANY_PTX_FILE)
+    config.CUDA_LOW_OCCUPANCY_WARNINGS = 0
+
+
+def _get_best_ptx_file(archs, max_compute_capability):
+    """
+    Determine of the available PTX files which one is
+    the most recent up to and including the device cc
+    """
+    filtered_archs = [x for x in archs if x[0] <= max_compute_capability]
+    if filtered_archs:
+        return max(filtered_archs, key=lambda y: y[0])
+    else:
+        return None
+
+
+def _get_ptx_file(path, prefix):
+    if "RAPIDS_NO_INITIALIZE" in os.environ:
+        # cc=60 ptx is always built
+        cc = int(os.environ.get("STRINGS_UDF_CC", "60"))
+    else:
+        from numba import cuda
+
+        dev = cuda.get_current_device()
+
+        # Load the highest compute capability file available that is less than
+        # the current device's.
+        cc = int("".join(str(x) for x in dev.compute_capability))
+    files = glob.glob(os.path.join(path, f"{prefix}*.ptx"))
+    if len(files) == 0:
+        raise RuntimeError(f"Missing PTX files for cc={cc}")
+    regular_sms = []
+
+    for f in files:
+        file_name = os.path.basename(f)
+        sm_number = file_name.rstrip(".ptx").lstrip(prefix)
+        if sm_number.endswith("a"):
+            processed_sm_number = int(sm_number.rstrip("a"))
+            if processed_sm_number == cc:
+                return f
+        else:
+            regular_sms.append((int(sm_number), f))
+
+    regular_result = None
+
+    if regular_sms:
+        regular_result = _get_best_ptx_file(regular_sms, cc)
+
+    if regular_result is None:
+        raise RuntimeError(
+            "This cuDF installation is missing the necessary PTX "
+            f"files that are <={cc}."
+        )
+    else:
+        return regular_result[1]
+
+
+def _setup_numba_linker(path):
+    try:
+        # ptxcompiler will not be present for cuda 12+
+        from ptxcompiler.patch import NO_DRIVER, safe_get_versions
+
+        versions = safe_get_versions()
+        if versions != NO_DRIVER:
+            driver_version, runtime_version = versions
+            ptx_toolkit_version = _get_cuda_version_from_ptx_file(path)
+            maybe_patch_numba_linker(
+                driver_version, runtime_version, ptx_toolkit_version
+            )
+    except ImportError:
+        pass
+
+
+def maybe_patch_numba_linker(
+    driver_version, runtime_version, ptx_toolkit_version
+):
+    # Numba thinks cubinlinker is only needed if the driver is older than
+    # the ctk, but when PTX files are present, it might also need to patch
```

Comment:
Just wanting to be precise in our comments here. Do we specifically mean the runtime rather than the “toolkit”?
```suggestion
    # the CUDA runtime, but when PTX files are present, it might also need to patch
```

Reply:
The toolkit component of concern is NVVM. We only use the runtime to check what version NVVM is (assuming if we find the runtime, that it's the same version as NVVM) because NVVM provides no way to check what version it is. I find it less confusing to refer to "the toolkit" as it's the collection of components including the runtime and NVVM, amongst other things.